### ISSUE TITLE
Use provided AuthenticationProvider automagically

### ISSUE THREAD

**Issue by chrylis** [spring-projects/spring-boot#2544]
Currently, replacing the default in-memory `AuthenticationProvider` with a custom one requires a substantial amount of non-obvious configuration. An `AuthenticationProvider` bean (perhaps exactly one) defined in the context should suppress the creation of the in-memory provider and be registered in its place.

**Comment by nomack84** [comment]
I think it will be harder to create an `AuthenticationProvider` bean than configuring the default in-memory provider. Perhaps a different choice could be defining the in-memory user in a properties configuration file, which I thinks it's already the done.

**Comment by rwinch** [comment]
I don't think this is a Boot feature. If implemented, it is something that belongs in Spring Security. 
If you are interested in this feature, I encourage you to vote on [SEC-2869|(https://jira.spring.io/browse/SEC-2869]

> Currently, replacing the default in-memory AuthenticationProvider with a custom one requires a substantial amount of non-obvious configuration.

Can you elaborate on why the configuration is any less obvious than providing an AuthenticationProvider Bean? You know you need to provide an AuthenticationProvider bean due to documentation, likewise you know you can use the builder due to documentation.

> An AuthenticationProvider bean (perhaps exactly one) defined in the context should suppress the creation of the in-memory provider and be registered in its place.

Spring Security (no Boot necessary) already allows you configure authentication by providing an AuthenticationManager as a bean. However, as @nomack84 points out, this is a bit more verbose than using the builder.

For example:

```java
@Bean
public AuthenticationManager authenticationManager(AuthenticationProvider provider) {
    return new ProviderManager(Arrays.asList(provider));
}
```

I acknowledge this is more verbose than providing just an AuthenticationProvider, but not by much so this isn't really a high priority at the moment.

> Perhaps a different choice could be defining the in-memory user in a properties configuration file, which I thinks it's already done.

Yes you can easily create an InMemoryUserDetailsManager from a Properties object using something like:

```java
        @Bean
        public UserDetailsService userDetailsService(@Value("classpath:users.properties") Resource users) {
            Properties properties = new Properties();
            properties.load(users.getInputStream());
            return new InMemoryUserDetailsManager(properties);
        }
```

**users.properties**
```properties
rob=robspassword,ROLE_A
luke=lukespassword,ROLE_A,ROLE_B
```

**Comment by rwinch** [comment]
Closing in favor of [SEC-2869](https://jira.spring.io/browse/SEC-2869)

**Comment by chrylis** [comment]
> I think it will be harder to create an AuthenticationProvider bean than configuring the default in-memory provider.

This misses the point: I want an `AuthenticationProvider` that implements a completely different strategy, but I would still be happy with the rest of the Spring Boot manager setup. 

> Can you elaborate on why the configuration is any less obvious than providing an AuthenticationProvider Bean?

I'm saying that providing an `AuthenticationProvider` should be *sufficient*. Currently, if I provide an `AuthenticationProvider`, it gets added *in addition* to the `InMemoryAuthenticationProvider`, and there's no way to suppress the latter without overriding the entire `AuthenticationManager` bean.

> Spring Security (no Boot necessary) already allows you configure authentication by providing an AuthenticationManager as a bean. However, as @nomack84 points out, this is a bit more verbose than using the builder.

I'm aware, but I'm quite satisfied with the arrangements that Spring Boot makes *except* that I want my own strategy applied in place of the `DaoAuthenticationProvider` that it adds by default. Easy overrides for the built-in provider strategies are available, but not for a custom strategy.

**Comment by rwinch** [comment]
> This misses the point: I want an AuthenticationProvider that implements a completely different strategy, but I would still be happy with the rest of the Spring Boot manager setup.

The following should disable Boot's default Authentication. Does this work for your needs?

```java
@Autowired
public void configureGlobal(AuthenticationManagerBuilder auth) {
  auth
    .authenticationProvider(myAuthenticationProvider());
}
```

> I'm aware, but I'm quite satisfied with the arrangements that Spring Boot makes except that I want my own strategy applied in place of the DaoAuthenticationProvider that it adds by default. Easy overrides for the built-in provider strategies are available, but not for a custom strategy.

If you provide your own custom AuthenticationManager, it will replace Boot's default DaoAuthenticationProvider.

**Comment by chrylis** [comment]
> The following should disable Boot's default Authentication.

I'm having a little trouble figuring out exactly how to make this trigger; I'm getting non-unique-order errors. I tried using that exact call inside an `@Override configure(AuthenticationManagerBuilder auth)`, and that appended my provider, leaving the default in-place. Since Boot's usual idiom is "provide your own bean to replace the default one", I expected that providing an `AuthenticationProvider` bean should just substitute for the `DaoAuthenticationProvider`.

> If you provide your own custom AuthenticationManager, it will replace Boot's default DaoAuthenticationProvider.

That's what I'm doing now, but that's replacing more of the infrastructure than I expected to have to; for example, providing my own `DataSource` is sufficient to get Spring Boot to plug it into JDBC-backed and JPA configurations.

**Comment by chrylis** [comment]
I have the `configureGlobal` working, with the caveat that it requires a magic number in the `@Order` annotation. Still an inconsistency with the way that most other bean replacement works in Boot.

**Comment by rwinch** [comment]
@chrylis Thanks for the response. Can you create a separate Boot issue with your specific issue (so I can reproduce it) and ensure to mention me on it? I'd like to see what sort of problems you are having.

**Comment by chrylis** [comment]
I'll try to isolate this out. I'm nearly certain that it's a Boot issue and not SEC, and especially something that would need to be addressed before 4.0.0.RELEASE.

**Comment by rwinch** [comment]
@chrylis Thanks for your willingness to provide a sample.

**Comment by chrylis** [comment]
https://github.com/chrylis/spring-boot-2544

**Comment by chrylis** [comment]
The Spring Boot code in question is in `BootDefaultingAuthenticationConfigurerAdapter`, which unconditionally uses the `inMemoryAuthentication().withUser()` builder. I would like this behavior to be `@ConditionalOnMissingBean(AuthenticationProvider)`.

**Comment by philwebb** [comment]
@rwinch Should this one be re-opened?

**Comment by rwinch** [comment]
**tl;dr** @chrylis - Can you please create a new ticket with the  `@Order` issue and an example?

@chrylis Thank you for the response.

Sorry I was not very clear. The request for an enhancement (i.e. only provide an AuthenticationProvider to override the defaults) is actually a request for Spring Security and is logged as [SEC-2869](https://jira.spring.io/browse/SEC-2869). When Spring Security resolves this, you will be able to provide an AuthenticationProvider Bean and Spring Security (including a Spring Boot application) will use that AuthenticationProvider.

Since that request is not a Boot feature, I have closed this ticket in favor of the previously mentioned Spring Security ticket.

When I said:

> Can you create a separate Boot issue with your specific issue (so I can reproduce it) and ensure to mention me on it? I'd like to see what sort of problems you are having.

I was referring to your comment about:

> I have the configureGlobal working, with the caveat that it requires a magic number in the @Order annotation.

The repository you provided appears to be illustrating [SEC-2869](https://jira.spring.io/browse/SEC-2869) rather than the `@Order` issue you mentioned. 

Can you please create a new ticket with the  `@Order` issue and an example?

**Comment by chrylis** [comment]
Sorry, I forgot which issue specifically I was trying to demonstrate. Regarding the distinction between SEC-2869 and this issue, it still seems that it belongs to Boot more than the core Security code. Setting up a `ProviderManager` given an existing `AuthenticationProvider` isn't difficult; I'm just wanting Boot to use that path in its autoconfiguration instead of unconditionally using the DSL to build a DAO provider.

If I'm able to revisit and figure out the `@Order` issue, then I'll try to put that up as well.

**Comment by rwinch** [comment]
@chrylis Thank you for your feedback.

> I'm just wanting Boot to use that path in its autoconfiguration instead of unconditionally using the DSL to build a DAO provider.

Under the covers [Boot is using Spring Security's AuthenticationConfiguration.getAuthenticationManager()](https://github.com/spring-projects/spring-boot/blob/v1.2.2.RELEASE/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/security/AuthenticationManagerConfiguration.java#L74) to determine what the AuthenticationManager to use.

If Boot tries to provide the AuthenticationManager based on the AuthenticationProvider, there will be a cycle between creating an AuthenticationManager in Boot and Spring Security's getAuthenticationManager() method since getAuthenticationManager() falls back to trying to resolve an AuthenticationManager Bean. 

In short, the feature needs to be in Security, but will also allow Boot users to leverage the feature as well.

**Comment by chrylis** [comment]
But isn't `BootDefaultingAuthenticationConfigurerAdapter` the one configuring that `AuthenticationManager` in the first place using the builder DSL?

**Comment by rwinch** [comment]
Hmm...breaking into its own Adapter would likely work, but I'm still interested in your goal.

Is there a *functional* reason that you feel it needs to be in Boot? There is a clear functional benefit to place it it in Security since the feature can be used by Boot users and non Boot users.

**Comment by chrylis** [comment]
Not if registering the `AuthenticationProvider` bean with the Spring context would be sufficient for a Boot app to get the default security configuration except with the custom provider. It seems, though, that the Boot starter is the component that wires up the security configuration, and it's where the behavior in question is.

**Comment by rwinch** [comment]
> Not if registering the AuthenticationProvider bean with the Spring context would be sufficient for a Boot app to get the default security configuration except with the custom provider

The changes SEC-2869 will implement will allow the sample project you submitted to work in a Boot or non Boot application.

**Comment by chrylis** [comment]
> The changes SEC-2869 will implement will allow the sample project you submitted to work in a Boot or non Boot application. Given that, I will leave this issue closed.

Looking forward to it.
